KAFKA-3186: KIP-50: Move Authorizer to a separate package

SinghAsDev opened the pull request.

harshach commented:
@Parth-Brahmbhatt  can you review this.

gwenshap commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -52,9 +55,13 @@ object AclCommand {
-        println(s"Error while executing ACL command: ${e.getMessage}")
-        println(Utils.stackTrace(e))
-        System.exit(-1)
+        if (inTestMode) {
```
I would much prefer if we would avoid this by calling addACL(opts) directly rather than main.
In general, both the extra switch for testing and unit-tests calling main() method are not recommended if they can at all be avoided, and I think in this case you can easily avoid it.

Parth-Brahmbhatt replied:
+1

SinghAsDev replied:
Sure, will update PR.

Parth-Brahmbhatt commented:
I am not really sure if we should restrict SimpleACLAuthorizer to only supporting UserType. Any reason it wont be able to support any other custom type that user has added if they decide to add a custom authentication layer?

SinghAsDev commented:
@Parth-Brahmbhatt the PR is not restricting SimpleAclAuthorizer to support only UserType, rather we are just exposing to authorizer what principal types authorizer's implementation supports. SimplaAclAuthorizer happens to only support "User" type as of now. A custom implementation can choose to support what ever types it has to. Exposing supported principal types at authorizer level, enables AclCommands or similar tools to check for user errors while specifying principal types.

Parth-Brahmbhatt commented:
@SinghAsDev I am not sure why SimpleAclAuthorizer would only support User type currently. Wouldn't it work with any other type as well even in its current state?

SinghAsDev commented:
@Parth-Brahmbhatt sure it will, however there is no way a user adding acls will ever know that principal types he/she is adding is even correct. For instance, KAFKA-3097.

Parth-Brahmbhatt commented:
I guess that's the trade off between stronger schema requirement for validation vs flexibility of supporting any open type. Originally the principal types were case insensitive and we changed it to case sensitive for consistency as topics in kafka are case sensitive. My preference would still be to support flexibility over schema validation. We can add better help messages in ACLCommand to avoid errors like KAFKA-3097. 

If we still chose to go with this PR, I recommend to at least make the supported type a configurable property so a 3rd party authentication scheme does not have to extend SimpleAclAuthorizer just to override the one method.

SinghAsDev commented:
Allowing pluggable authorization is the way to support flexibility, right.
Supported principal types is really something that varies with authorizer
implementation. I do not see why we need one more configuration, when third
party can just provide the info via authorizer interface. Makes sense?

On Thu, Feb 4, 2016 at 10:47 AM, Parth Brahmbhatt notifications@github.com
wrote:

> I guess that's the trade off between stronger schema requirement for
> validation vs flexibility of supporting any open type. Originally the
> principal types were case insensitive and we changed it to case sensitive
> for consistency as topics in kafka are case sensitive. My preference would
> still be to support flexibility over schema validation. We can add better
> help messages in ACLCommand to avoid errors like KAFKA-3097.
> 
> If we still chose to go with this PR, I recommend to at least make the
> supported type a configurable property so a 3rd party authentication scheme
> does not have to extend SimpleAclAuthorizer just to override the one
> method.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/apache/kafka/pull/861#issuecomment-179996162.

## 

Regards,
Ashish

Parth-Brahmbhatt commented:
Lets take the scenario of someone wanting custom authentication but default authorization(SimpleAclAuthorizer):

Current scenario:
- No extra work needed but no help message displayed to the user about newly added types.

After this patch:
- Extend SimpleAclAuthorizer.
- Override a single method getSupportedPrincipalTypes to list the custom principalType that their custom authentication added.
- General packaging of this new custom authorizer with all the other classes and configuration changes to make this new class the authorizer.

If we add config for supported Types:
- Only the config for supported type has to change.

Let me know if I am wrong in the above evaluation. In addition, if we decide that we want to add stronger validations as recommended in this PR, I would recommend that we add actual validation in SimpleAclAuthorizer. In the current PR someone trying to add the ACLs using CLI will get an error and someone adding Acls programmatically won't.

SinghAsDev commented:
@Parth-Brahmbhatt please find my response inline.

> Current scenario:
> No extra work needed but no help message displayed to the user about newly added types.
> After this patch:

This is what we want to fix. This will help us avoid bunch of user group/ JIRA questions/complains on why Kafka authorization is not working for them.

> Extend SimpleAclAuthorizer.
> Override a single method getSupportedPrincipalTypes to list the custom principalType that their custom authentication added.
> General packaging of this new custom authorizer with all the other classes and configuration changes to make this new class the authorizer.

This is how the community agreed on supporting pluggable authorization, build custom authorizer implementation and package it. This PR does not change current behavior of `SimpleAclAuthorizer`. This is how third party authorization implementations would work even now, right?

> If we add config for supported Types:
> Only the config for supported type has to change.

Yet another config. We should try to avoid any new configs, however if multiple people are in favor of this we can go ahead with this. @gwenshap @ijuma what do you guys think?

> I would recommend that we add actual validation in SimpleAclAuthorizer. In the current PR someone trying to add the ACLs using CLI will get an error and someone adding Acls programmatically won't.

The reason this PR has the checks in CLI is to fail fast. I see your point on programatic access of `SimpleAclAuthorizer`. However, we need to define the expectation of Authorizer interface. We can either say that Authorizer interface, `addAcls` and `removeAcls`, will fail for improper Acls or we can say it does not check. When we are exposing `getSupportedPrincipalTypes`, it is only fair to assume that onus is on clients/callers to check if they are passing proper ACLs in create/remove acls commands.

SinghAsDev commented:
@gwenshap could you take a look when you get a chance, thanks.

SinghAsDev commented:
@gwenshap ping for review.

granthenke commented on core/src/main/scala/kafka/security/auth/Authorizer.scala:
```diff
@@ -83,6 +83,13 @@ trait Authorizer extends Configurable {
+   * @return set containing only "User" as principal type if Authorizer implementation does not provide this information,
+   *         else set of principal types supported by the configured authorizer.
+   */
+  def getSupportedPrincipalTypes(): Set[String] = Set(KafkaPrincipal.USER_TYPE)
```
Since this is just the interface, it probably shouldn't have any implementation.

SinghAsDev replied:
This is for backwards compatibility. Existing third party implementations should be able to work without making any changes.

ijuma replied:
Note that this is only source compatible, but not binary compatible.

ijuma commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -91,10 +93,12 @@ object AclCommand {
   }
 
-  private def removeAcl(opts: AclCommandOptions) {
+  def removeAcl(opts: AclCommandOptions) {
```
Why are these two methods public now?

SinghAsDev replied:
For testing. Can't rely on `AclCommand.main` as it calls System.exit on getting an exception. `testInvalidPrincipalTypeWhileRemovingAcls` should probably use `removeAcl`.

ijuma replied:
Can't you just do `private[admin]` or something like that?

SinghAsDev replied:
That works, thanks!

ijuma commented:
Since this changes a public interface and it seems like there is no consensus on whether we should do it or not, doesn't that imply it needs a KIP?

SinghAsDev commented:
I can do a short KIP. The point @Parth-Brahmbhatt and I were discussing is should we have principal types specified via a config or via interface. We can discuss that over KIP.

SinghAsDev commented:
@ijuma just added KIP-50, https://cwiki.apache.org/confluence/display/KAFKA/KIP-50+-+Enhance+Authorizer+interface+to+be+aware+of+supported+Principal+Types.

@Parth-Brahmbhatt I have added your opinion on having a separate config option in the KIP. Based on how the KIP discussion goes, I will make changes accordingly. Do you mind reviewing the KIP.

Parth-Brahmbhatt commented:
@SinghAsDev The KIP looks good to me. You can probably add that once this method is added we expect the Client of the authorizer to do the validation on principal types and the authorizer will still not do any validation by it self. As an alternative we can add the validation at Authorizer level.

SinghAsDev commented:
@Parth-Brahmbhatt good point. Will add that shortly.

ijuma commented on core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala:
```diff
@@ -227,6 +227,10 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
   }
 
+  override def getSupportedPrincipalTypes(): Set[String] = {
+    Set(KafkaPrincipal.USER_TYPE)
```
Why are you overriding this method since the implementation is the same as in the interface?

ijuma commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -124,6 +128,17 @@ object AclCommand {
   }
 
+  def validatePrincipalTypes(resourceToAcl: Map[Resource, Set[Acl]], validPrincipalTypes: Set[String]) = {
+    for((_, acls) <- resourceToAcl) {
```
`resourceToAcl.values` iterates over the values, which is what you want.

ijuma replied:
Why are we receiving `resourceToAcl` if we only care about the ACLs? Maybe you want to use that information to provide better error messages? If not, then we should change what we receive to be `Set[Acl]` or something.

ijuma commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -124,6 +128,17 @@ object AclCommand {
     }
   }
 
+  def validatePrincipalTypes(resourceToAcl: Map[Resource, Set[Acl]], validPrincipalTypes: Set[String]) = {
```
This method has a bunch of cases where the space after `for`/`if` are missing.

ijuma commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -93,10 +95,12 @@ object AclCommand {
     withAuthorizer(opts) { authorizer =>
       val resourceToAcl = getResourceToAcls(opts)
 
+      validatePrincipalTypes(resourceToAcl, authorizer.getSupportedPrincipalTypes())
```
I wonder if we should do the validation as part of `getResourceToAcls` instead of adding it in multiple places.

Commit:
KIP-50
1. Move auth related classes to separate Java package.
2. Add description() to authorizer interface.
3. Move SimpleAclAuthorizer specific methods from Acl and other
authorizer related classes to SimpleAclAuthorizer.
4. Reorganize tests so that authorizer does not have to depend on
clients.
5. Create Configurable interface, to avoid clients dependency in
authorizer.
6. A few classes have been taken from, PR #1005. We can move common code to
separate PR, or just let the later patch rebase.

granthenke commented on clients/src/main/java/org/apache/kafka/common/errors/InvalidAclException.java:
```diff
@@ -6,33 +6,30 @@
+package org.apache.kafka.common.errors;
 
-class OperationTest extends JUnitSuite {
+/**
```
I am not sure why Git is marking this as a Rename. It makes review difficult. @SinghAsDev Can you see if this can be fixed? Perhaps squashing the commits would help?

SinghAsDev replied:
Yea, it is weird. Will try squashing commits.

SinghAsDev replied:
Even squashing commits did not help.

granthenke commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -29,17 +29,17 @@ object AclCommand {
 
     if (opts.options.has(opts.helpOpt))
-      CommandLineUtils.printUsageAndDie(opts.parser, "Usage:")
+      CommandLineUtils.printUsageAndDie(opts.parser, description(opts))
```
We still may want the "Usage:" string output. Perhaps the description should come after.

granthenke commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -76,6 +76,11 @@ object AclCommand {
   }
 
+  private def description(opts: AclCommandOptions): String = {
+    withAuthorizer(opts) { authorizer => return authorizer.description()}
```
This syntax is a bit unusual. Especially using a return and also the "" at the end of the method. I think this could be cleaned up by changing the signature of `withAuthorizer` to return the type of the passed function.

From:

``` scala
def withAuthorizer(opts: AclCommandOptions)(f: Authorizer => Unit) {
```

To:

```
def withAuthorizer[T](opts: AclCommandOptions)(f: Authorizer => T): T = {
```

SinghAsDev replied:
Thanks for the cleanup suggestion!

granthenke commented on clients/src/main/java/org/apache/kafka/common/errors/InvalidAclException.java:
```diff
@@ -6,33 +6,30 @@
+/**
+ * Throw when an invalid Acl is being added or removed.
+ */
+public class InvalidAclException extends ApiException {
```
Would it make sense to have a top level "AuthorizerException" that the other exceptions can inherit from? That way if there is an implementation specific exception that doesn't match the expected ones. Implementations can still throw a generic exception that can still be caught and handled appropriately.

granthenke replied:
Actually we could probably use "AuthorizationException"

granthenke commented on clients/src/main/java/org/apache/kafka/common/security/auth/Authorizer.java:
```diff
@@ -0,0 +1,119 @@
+     * @param resource  Resource the client is trying to access.
+     * @return
+     *
+     * @throws org.apache.kafka.common.errors.InvalidResourceException if resource does not exist
```
I am not sure InvalidResourceException should be thrown in the resource doesn't exist. It may be better to just return false on authorize. I think that is what the default authorizer does.

SinghAsDev replied:
I guess the description can be improved. If it is something like `topic:t1` and `t1` topic does not exist, then it makes sense to return false. However, if it is something like `bla:bla` wouldn't it be better to thrown the exception?

ewencp replied:
I think just say "if the resource type does not exist" to constrain it the . This might need more clarification in Javadocs for users implementing this since "resource does not exist" seems to require returning false, but `InvalidResourceException` really means invalid resource _type_ (but someone could easily interpret "invalid resource" to include non-existant but correctly-defined resources).

granthenke commented on clients/src/main/java/org/apache/kafka/common/security/auth/Authorizer.java:
```diff
@@ -0,0 +1,119 @@
+     * @return
+     *
+     * @throws org.apache.kafka.common.errors.InvalidResourceException if resource does not exist
+     * @throws org.apache.kafka.common.errors.InvalidOperationException if requested operation is not
```
This could be handled by returning false too, perhaps an exception is more appropriate here though.

SinghAsDev replied:
Yea, I too think having exception is better here.

granthenke commented on clients/src/main/java/org/apache/kafka/common/security/auth/Authorizer.java:
```diff
@@ -0,0 +1,119 @@
+     * @param resource the resource to which these acls should be attached.
+     *
+     * @throws org.apache.kafka.common.errors.AuthorizationException if not authorized to add acls for the resource
+     * @throws org.apache.kafka.common.errors.InvalidResourceException if resource does not exist
```
I think the default implementation lets you set ACLs for resources that don't exist yet.

SinghAsDev replied:
Yea, this is again for something like `bla:bla`. Let me know what is your opinion on that.

granthenke commented on clients/src/main/java/org/apache/kafka/common/security/auth/Authorizer.java:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @throws org.apache.kafka.common.errors.AuthorizationException if not authorized to add acls for the resource
+     * @throws org.apache.kafka.common.errors.InvalidResourceException if resource does not exist
+     * @throws org.apache.kafka.common.errors.InvalidAclException if an invalid acl is being added
```
Just calling out that because this is a void function, we can't report back what ACLs have and have not been added. When an exception is thrown, the user can't be sure if some ACLs were added, none were added, or all were added. They also can't tell which ACL caused the exception.

SinghAsDev replied:
That is a good point. Not sure what is the best solution here, but user can always list acls to see what got added and what failed.

granthenke commented on clients/src/main/java/org/apache/kafka/common/security/auth/Authorizer.java:
```diff
@@ -0,0 +1,119 @@
+     *
+     * @param acls     set of acls to be removed.
+     * @param resource resource from which the acls should be removed.
+     * @return true if some acl got removed, false if no acl was removed.
```
Same thoughts as above:

> return true if some acl got removed, false if no acl was removed.

There is no feedback on which ones got removed and which didn't.

SinghAsDev replied:
Same as above.

> That is a good point. Not sure what is the best solution here, but user can always list acls to see what got removed and what failed.

SinghAsDev commented:
@harshach @ijuma @gwenshap pinging you guys for review here.

ewencp commented:
@SinghAsDev I think this needs notes added to `docs/upgrade.html` too since it includes breaking changes.

Commit:
Add note to upgrade docs to indicate how third party authorizers will be affected.

SinghAsDev commented:
@ewencp good point, added note to upgrade doc.

SinghAsDev commented:
@ewencp mind giving it another look?

ewencp commented on clients/src/main/java/org/apache/kafka/common/security/auth/Authorizer.java:
```diff
@@ -0,0 +1,119 @@
+     * @param session   The session being authenticated.
+     * @param operation Type of operation client is trying to perform on resource.
+     * @param resource  Resource the client is trying to access.
+     * @return
```
Missing specification of return value (which was missed in the KIP too...)

Commit:
Standardize and add missing java docs.

SinghAsDev commented:
@ewencp updated KIP's doc and PR to add missing java docs.

Commit:
Rebase over trunk.

ewencp commented:
@SinghAsDev I'm getting pretty consistent errors in some of the tests that all look something like this:

```
kafka.api.SslEndToEndAuthorizationTest > testNoConsumeAcl FAILED
    java.lang.AssertionError: expected acls Set(User:O=A client,CN=localhost has Allow permission for operations: Write from hosts: *, User:O=A client,CN=localhost has Allow permission for operations: Describe from hosts: *) but got Set(User:O=A client,CN=localhost has Allow permission for operations: Describe from hosts: *, User:O=A server,CN=localhost has Allow permission for operations: Read from hosts: *, User:O=A client,CN=localhost has Allow permission for operations: Write from hosts: *)
        at org.junit.Assert.fail(Assert.java:88)
        at kafka.utils.TestUtils$.waitUntilTrue(TestUtils.scala:768)
        at kafka.utils.TestUtils$.waitAndVerifyAcls(TestUtils.scala:1093)
        at kafka.api.EndToEndAuthorizationTest$$anonfun$testNoConsumeAcl$1.apply(EndToEndAuthorizationTest.scala:225)
        at kafka.api.EndToEndAuthorizationTest$$anonfun$testNoConsumeAcl$1.apply(EndToEndAuthorizationTest.scala:224)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
        at kafka.api.EndToEndAuthorizationTest$class.testNoConsumeAcl(EndToEndAuthorizationTest.scala:224)
        at kafka.api.SslEndToEndAuthorizationTest.testNoConsumeAcl(SslEndToEndAuthorizationTest.scala:24)
```

SinghAsDev commented:
@ewencp it passes each time with Java8, however frequently, but not always, fails on 1.7. I am looking into the test failures. Did you see the errors with Java 7 or 8?

ewencp commented:
@SinghAsDev JDK 7

Commit:
Fix test failures.

SinghAsDev commented:
@ewencp the test failures are fixed now. Build failed due to insufficient disk space on build machine. It works fine on my machine with Java7.

```
# There is insufficient memory for the Java Runtime Environment to continue.
# Cannot create GC thread. Out of system resources.
# An error report file with more information is saved as:
# /home/jenkins/.gradle/daemon/2.4-rc-2/hs_err_pid13065.log
```

ijuma commented on clients/src/main/java/org/apache/kafka/common/security/auth/Authorizer.java:
```diff
@@ -0,0 +1,119 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.common.security.auth;
```
Personally, I think it's not a good idea to add server-specific code in the `common` package. How about introducing a `server` package? It is weird to have that in the `clients` jar, but at least it's clear for users that this is not something used by both clients and servers.

Do we envisage any scenario where the `Authorizer` is used by clients?

ijuma replied:
It would be good to get more opinions on this.

SinghAsDev replied:
This has been discussed and agreed upon in KIP-50 discuss thread.

ijuma replied:
I'm aware of the discussion, I was in that thread. :) Also long quotes make it harder to review the code. As far as I can see, the discussion you quoted is talking about jars. My suggestion is with regards to package names.

ijuma replied:
Also, Jay said:

> 1. Anything which is just in the clients is in org.apache.clients under clients/
> 2. Anything which is in the server is kafka.\* which is under core/
> 3. Anything which is needed in both places (as it sounds like some
>    enums
>    for authorization are?) is in common which is under clients

Under that definition, the `authorizer` would be under `2`.

SinghAsDev replied:
`Authorizer` is used by third party libs and there is no strong reason for them to depend on Kafka core.

Apologies about the long excerpt, I will edit my previous comment to strip out the long excerpt part. Regarding introducing a `server` package, I do not have a scenario where `Authorizer` will be used in clients and as such I am fine with your suggestion.

ijuma replied:
OK, great. Let me send a message to the KIP discussion to get wider input on the package name.

SinghAsDev replied:
@ijuma it has been some time since you posted on discuss list and there is no feedback. Should we just go ahead and update the KIP and PR?

SinghAsDev replied:
@ijuma will appreciate some feedback here.

ijuma commented on clients/src/main/java/org/apache/kafka/common/security/auth/Authorizer.java:
```diff
@@ -0,0 +1,119 @@
+import java.util.Set;
+
+/**
+ * Top level interface that all plugable authorizer must implement. Kafka server will read "authorizer.class" config
```
`pluggable`

omkreddy commented:
This was done in KAFKA-8865
